common/cuda_hip/factorization/lu_kernels.hpp.inc

```diff
@@ -0,0 +1,206 @@
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+namespace kernel {
+
+
+template <typename ValueType, typename IndexType>
+__global__ __launch_bounds__(default_block_size) void initialize(
+    const IndexType* __restrict__ mtx_row_ptrs,
+    const IndexType* __restrict__ mtx_cols,
+    const ValueType* __restrict__ mtx_vals,
+    const IndexType* __restrict__ factor_row_ptrs,
+    const IndexType* __restrict__ factor_cols,
+    const IndexType* __restrict__ factor_storage_offsets,
+    const int32* __restrict__ factor_storage,
+    const int64* __restrict__ factor_row_descs,
+    ValueType* __restrict__ factor_vals, IndexType* __restrict__ diag_idxs,
+    size_type num_rows)
+{
+    const auto row = thread::get_subwarp_id_flat<config::warp_size>();
+    if (row >= num_rows) {
+        return;
+    }
+    const auto warp =
+        group::tiled_partition<config::warp_size>(group::this_thread_block());
+    // first zero out this row of the factor
+    const auto factor_begin = factor_row_ptrs[row];
+    const auto factor_end = factor_row_ptrs[row + 1];
+    const auto lane = static_cast<int>(warp.thread_rank());
+    for (auto nz = factor_begin + lane; nz < factor_end;
+         nz += config::warp_size) {
+        factor_vals[nz] = zero<ValueType>();
+    }
+    warp.sync();
+    // then fill in the values from mtx
+    gko::matrix::csr::device_sparsity_lookup<IndexType> lookup{
+        factor_row_ptrs, factor_cols,      factor_storage_offsets,
+        factor_storage,  factor_row_descs, row};
+    const auto row_begin = mtx_row_ptrs[row];
+    const auto row_end = mtx_row_ptrs[row + 1];
+    for (auto nz = row_begin + lane; nz < row_end; nz += config::warp_size) {
+        const auto col = mtx_cols[nz];
+        const auto val = mtx_vals[nz];
+        factor_vals[lookup.lookup_unsafe(col) + factor_begin] = val;
+    }
+    if (lane == 0) {
+        diag_idxs[row] = lookup.lookup_unsafe(row) + factor_begin;
+    }
+}
+
+
+template <typename ValueType, typename IndexType>
+__global__ __launch_bounds__(default_block_size) void factorize(
+    const IndexType* __restrict__ row_ptrs, const IndexType* __restrict__ cols,
+    const IndexType* __restrict__ storage_offsets,
+    const int32* __restrict__ storage, const int64* __restrict__ row_descs,
+    const IndexType* __restrict__ diag_idxs, ValueType* __restrict__ vals,
+    int* __restrict__ ready, int* __restrict__ block_counter,
+    size_type num_rows)
+{
+    __shared__ int threadblock_offset;
+    __shared__ int sh_ready[default_block_size / config::warp_size];
+    if (threadIdx.x == 0) {
+        threadblock_offset = atomic_add(block_counter, 1) * default_block_size;
+    }
```

Comment:
CUDA doesn't guarantee that new blocks are scheduled in a monotonic order (even though some NVIDIA libraries like cub rely on it), so just to be safe, we added this workaround. I'm also pretty annoyed with them for this behavior, but it might change due to more complex scheduling on multi-chiplet GPUs or so, I could imagine? Anyways, without this monotonicity, we could livelock by starving previous rows of compute time.